ngraph/core/reference/src/runtime/reference/pad.cpp

```diff
@@ -16,198 +16,293 @@
 
 #include "ngraph/runtime/reference/pad.hpp"
 
+#include "ngraph/axis_vector.hpp"
+#include "ngraph/check.hpp"
+#include "ngraph/coordinate_index.hpp"
+#include "ngraph/coordinate_transform.hpp"
+
+#include <cassert>
+
 namespace ngraph
 {
     namespace runtime
     {
-        namespace reference
+        namespace impl
         {
-            void pad(const char* data,
-                     const char* pad_value,
-                     char* out,
-                     const size_t elem_size,
-                     const Shape& data_shape,
-                     const Shape& out_shape,
-                     const CoordinateDiff& padding_below,
-                     const CoordinateDiff& padding_above,
-                     const op::PadMode pad_mode)
+            namespace
             {
-                Coordinate input_start(data_shape.size(), 0); // start at (0,0,...,0)
-                Coordinate input_end = out_shape; // end at (d'0,d'1,...,d'n), the outer corner of
-                                                  // the post-padding shape
-
-                Strides input_strides(data_shape.size(), 1);
-
-                AxisVector input_axis_order(data_shape.size());
-                for (size_t i = 0; i < data_shape.size(); i++)
+                template <typename T>
+                T clamp(T v, T lo, T hi)
                 {
-                    input_axis_order[i] = i;
+                    if (v < lo)
+                    {
+                        return lo;
+                    }
+                    if (v > hi)
+                    {
+                        return hi;
+                    }
+                    return v;
                 }
+                struct PadBase
```

Comment:
Why is it not a class with private or protected fields?

Reply:
IMHO: struct fill better here because:
1. class with protected members ale even more evil than struct (you have still some directly access to class members).
2. all Pad's structures are in one cpp file in unnamed namespace.